### PR TITLE
Add ability to specify commit of avalanchego in ci

### LINKED ISSUE #500: Update E2E Tests to Optionally run with AvalancheGo branch/commit instead of release
**Context and scope**
Currently the e2e tests that run in CI depend on an AvalancheGo release, which is installed in a specific location to be used in the e2e tests here: https://github.com/ava-labs/subnet-evm/blob/master/.github/workflows/lint-tests-release.yml#L66-L68

We should update this so that if a commit or branch name is specified in `scripts/versions.sh` here: https://github.com/ava-labs/subnet-evm/blob/master/scripts/versions.sh#L6 then it will checkout the relevant commit/branch of AvalancheGo and build it in order to run the e2e tests.

This will improve the CI workflow when there are breaking changes made to AvalancheGo that we would like to test in Subnet-EVM.

### PR DESCRIPTION
**Pull request by richardpringle**
## Why this should be merged

Resolves #500

## How this works

use the `AVALANCHE_VERSION` environment variable to specify a commit or branch name (or tag without a prebuilt binary).

## How this was tested

tested here for a `commit`:
https://github.com/ava-labs/subnet-evm/actions/runs/4512439095/jobs/7945887581

Screenshot 2023-03-24 at 10 54 24 AM

and here for a `branch`:
https://github.com/ava-labs/subnet-evm/actions/runs/4512412729/jobs/7945824005

Screenshot 2023-03-24 at 10 56 36 AM

## How is this documented

documentation for the whole process needs to be added

### CODE REVIEW

File: scripts/install_avalanchego_release.sh
Comment (darioush) [inline scripts/install_avalanchego_release.sh:53]: should we make this a shallow clone of the repository and avoid fetching the entire repository if the dir exists?

File: scripts/install_avalanchego_release.sh
Comment (darioush) [inline scripts/install_avalanchego_release.sh:36]: could we move this to the download part of the `if` or update the language so it says `checking for` instead of downloading?

Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:36]: you bet

Comment (richardpringle) [reply scripts/install_avalanchego_release.sh]: Screenshot 2023-03-22 at 3 21 00 PM

thoughts

File: scripts/install_avalanchego_release.sh
Comment (aaronbuchwald) [inline scripts/install_avalanchego_release.sh:32]: is the question why do we install this in a directory without the version as a suffix?

Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:32]: Yeah... made me think that it's being used elsewhere.

Comment (aaronbuchwald) [reply scripts/install_avalanchego_release.sh:32]: Currently we use this as a utility in local development to install a release of AvalancheGo, but we use it more frequently in CI to download a specific release that Subnet-EVM depends on to run its e2e tests.

In CI, we populate this environment variable here: https://github.com/ava-labs/subnet-evm/blob/master/.github/workflows/ci.yml#L77 and then pass the same variable in to run our ginkgo e2e tests here: https://github.com/ava-labs/subnet-evm/blob/master/.github/workflows/ci.yml#L83 (we could definitely clean this up a bit).

Just noticed we actually do something else that is weird here which is we use `BASE_DIR` instead of `BASEDIR` in one place in the script (on master) here: https://github.com/ava-labs/subnet-evm/blob/master/scripts/install_avalanchego_release.sh#L21.

Running this script locally now I'm running into the issue that instead of installing in `/tmp/` it's actually installing in the Subnet-EVM directory in the build path: `v1.9.11:-/tmp/avalanchego-release/avalanchego-v1.9.11`.

It looks like this is because the syntax `${var:-<alternative>}` checks if `var` is populated and uses it if so, but otherwise uses `<alternative>`. Since this change is not using a single variable, it seems to just use the whole string `v1.9.11:-/tmp/avalanchego-release/avalanchego-v1.9.11` instead of putting it in `/tmp`.

Comment (richardpringle) [reply scripts/install_avalanchego_release.sh]: Okay, take a look now. I left the commit as a fixup and will squash once I get the green light. Either we keep these changes or I can figure out a better way to deal with the branching logic.

File: scripts/install_avalanchego_release.sh
Comment (ceyonur) [inline scripts/install_avalanchego_release.sh:31]: You mean why it's not being used or why is not working? I don't think we use any `AVALANCHE_VERSION=xxx` as an env var in CI currently. Subnet-EVM (or any other rpcchainVM) cannot work with all AvalancheGO versions. There is a protocl version handshake, and if AvalancheGo bumps that protocol version and older subnet-evm cannot work with it. Similarly a recent Subnet-EVM (with newer protocol version) won't work with an older AvalancheGo version. e2e tests takes around ~5 min. It's probably best to use latest possible avalanchego for Pull request CI (to reduce test time). However we can include "all supported" avalanchego versions in a release-CI.

File: .github/workflows/ci.yml
Comment (aaronbuchwald) [inline .github/workflows/ci.yml:77]: Can we align all of these `BASEDIR` and `BASE_DIR` to be the same? The previous bug was that they were ever different as the intended behavior is to allow someone to populate the environment variable and use the default if not populated.

File: scripts/install_avalanchego_release.sh
Comment (aaronbuchwald) [reply scripts/install_avalanchego_release.sh:35]: I agree with this, I'd prefer to just use a single `AVALANCHEGO_BUILD_PATH` and either build in that location or unpack the downloaded release into that location rather than adding a separate versioned path.

File: .github/workflows/ci.yml
Comment (richardpringle) [reply .github/workflows/ci.yml:77]: I don't mind doing that, but it looks like it was done on purpose as the pattern repeats itself in a couple of places, `AVALANCHE_VERSION` vs `AVALANCHE_GO_VERSION` for instance. `BASEDIR` is only meant to be used inside the script.

If I assume that using a "private" variable name like that is justified, I think it would be more pragmatic to check if the user has defined `BASEDIR` instead of `BASE_DIR` and error out with an explanation. I can do the same for `AVALANCHE_GO_VERSION` vs `AVALANCHE_VERSION`

File: scripts/install_avalanchego_release.sh
Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:31]: >You mean why it's not being used or why is not working?

@ceyonur, thanks for answering! It actually is being used in the last step for the e2e tests in CI. I can actually remove the question now.

Comment (aaronbuchwald) [reply scripts/install_avalanchego_release.sh:31]: This is an interesting idea. As @ceyonur mentioned AvalancheGo and Subnet-EVM can only talk to each other if they are on the same RPCChainVMProtocolVersion.

AvalancheGo documents this for all releases here: https://github.com/ava-labs/avalanchego/blob/master/version/compatibility.json
Subnet-EVM documents this for all releases here: https://github.com/ava-labs/subnet-evm/blob/master/compatibility.json

I think the script (and whats in scope for this PR) should make sure that we can install any tag or commit-ish of AvalancheGo. In the future, your suggestion of moving this to a generally available install script in AvalancheGo and even a supported GitHub action would be really nice too.

As for what we run in CI, we could definitely add a check to CI that runs the Subnet-EVM e2e tests with every version of AvalancheGo that should be supported, which could just use this script (and actually would not depend on this change if we just wanted to run with official AvalancheGo releases). Although, I'd prefer this as a cronjob or manual GitHub action because it seems like overkill to run frequently given both the rpcchainvm server and client are in AvalancheGo, so I think it would make more sense to run this to confirm prior to releases or at a weekly cadence.

File: scripts/install_avalanchego_release.sh
Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:37]: You're absolutely right for commits and binary-releases, but this doesn't apply for branches, which was specified in the [original issue](https://github.com/ava-labs/subnet-evm/issues/500).

>We should update this so that if a commit or **branch name**

(emphasis mine)

File: .github/workflows/ci.yml
Comment (aaronbuchwald) [reply .github/workflows/ci.yml:77]: In this case, I think both should be public environment variables that can be easily overridden by the user and I'd prefer to be consistent in using the same environment variable in the script that the user needs to override rather than renaming in the background.

I think all of the renaming is confusing when looking at the script and wanting to know how to override a variable or where it comes from.

If we want to go back to having public vs. private env variables, I think we should be explicit and follow the convention of using lowercase for private env variables.

File: scripts/install_avalanchego_release.sh
Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:82]: See my response [here](https://github.com/ava-labs/subnet-evm/pull/589/files#r1146294527)

File: scripts/install_avalanchego_release.sh
Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:35]: @aaronbuchwald, why do you prefer a single build path? 

@ceyonur, the reason I did this was actually to avoid further nesting of logic, making the whole script harder to read. Not the best reason, but a reason, none-the-less.

I'm now thinking that `AVALANCHE_VERSION` is too overloaded and that it would be better to separate things into `AVALANCHE_RELEASE_VERSION`, `AVALANCHE_COMMIT`, and `AVALANCHE_BRANCH` and the logic for each variant is completely different (branches need to pulled in every time).

@aaronbuchwald, I also think it makes a lot more sense to have versioned paths, the same way that `cargo` stores dependencies for instance. I can make changes such that we don't rebuild if an existing commit or release-tagged version already exists, and only rebuild for a specified branch. 

Thoughts?

Comment (aaronbuchwald) [reply scripts/install_avalanchego_release.sh]: Sounds good to me 👍

File: scripts/versions.sh
Comment (aaronbuchwald) [inline scripts/versions.sh:7]:
```suggestion
AVALANCHEGO_VERSION=${AVALANCHEGO_VERSION:-'v1'.9.11'}
```
These are the same thing, can we make them one variable

File: scripts/install_avalanchego_release.sh
Comment (aaronbuchwald) [inline scripts/install_avalanchego_release.sh:35]: Nice

File: scripts/install_avalanchego_release.sh
Comment (aaronbuchwald) [inline scripts/install_avalanchego_release.sh:62]:
```suggestion
    # else the version is a git commitish (or it's invalid)
```

File: scripts/install_avalanchego_release.sh
Comment (aaronbuchwald) [inline scripts/install_avalanchego_release.sh:23]: Would it be better to use a single variable for this instead of `VERSION` and `AVALANCHEGO_VERSION` ?

File: scripts/install_avalanchego_release.sh
Comment (aaronbuchwald) [inline scripts/install_avalanchego_release.sh:33]: This seems to fail locally unless you specify `AVALANCHEGO_BUILD_PATH` (this passes CI since it manually populated this env var)

If we are going to use it in `BUILD_DIR` here, then we should update this to ensure that if it's not populated it defaults to a reasonable location like the previous behavior defaulting to use the already populated `BASEDIR`

File: scripts/install_avalanchego_release.sh
Comment (aaronbuchwald) [inline scripts/install_avalanchego_release.sh:104]: Hmm at first I thought that the below code where it builds iff the build dir does not exist may run into issues for branches if there were newly pushed changes, but I see that this ensures that the build dir is based off of the commit, so that's not an issue.

File: scripts/versions.sh
Comment (richardpringle) [reply scripts/versions.sh:7]: Yeah, I was a little worried about `AVALANCHE_VERSION` being used in other scripts though. My change makes this backwards compatible, so you can change the other scripts one at a time and make sure you don't break anything. I think that would be be better as a follow up PR so the context is a little more clear. 

So how about I do that immediately after this is merged?

File: scripts/install_avalanchego_release.sh
Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:35]: 😎  (why I can't do this as a reaction is beyond me)

File: scripts/install_avalanchego_release.sh
Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:23]: Yeah, I didn't really mean to move this. I kept `VERSION` because it makes the next couple of lines shorter. Happy to change it though

File: scripts/install_avalanchego_release.sh
Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:33]: Agreed! Will change

File: scripts/install_avalanchego_release.sh
Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:104]: I'm brilliant, what can I say, 😉

File: scripts/versions.sh
Comment (aaronbuchwald) [reply scripts/versions.sh:7]: Sounds good to me

File: scripts/install_avalanchego_release.sh
Comment (aaronbuchwald) [inline scripts/install_avalanchego_release.sh:78]: I think this needs to be `AVALANCHEGO_VERSION` same for the other two places below

**Review by aaronbuchwald — APPROVED** [review]
LGTM

File: scripts/install_avalanchego_release.sh
Comment (darioush) [inline scripts/install_avalanchego_release.sh:38]: should we drop `v`?

File: scripts/install_avalanchego_release.sh
Comment (darioush) [inline scripts/install_avalanchego_release.sh:76]: I still think we shouldn't fetch or clone the entire repository to get a specific branch or commit.

File: scripts/install_avalanchego_release.sh
Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:38]: I don't think it's worth the change. Every time I make a change, I have to run CI three different times to make sure everything is working as expected (one for tagged version, one for branch, and one for commit). While I know that dropping a "verbose" option won't change any functionality, if I don't run those tests again, the screen shots and linked actions won't actually contain a commit in this branch. 

Long winded way of saying, "if we want to make that change, it should happen in another PR".

File: scripts/install_avalanchego_release.sh
Comment (richardpringle) [reply scripts/install_avalanchego_release.sh:76]: What do you mean by "fetch or clone the entire repository"? To my knowledge, you can't partially fetch a repository (unless you have git submodules or you're using LFS or something like that). 

From the git docs:
>When no remote is specified, by default the origin remote will be used, unless there’s an upstream branch configured for the current branch.

Regardless, `fetch` will grab everything we need here, then we always checkout a detached-head by prefixing branch names with `origin/<branch_name>` or a commit directly.

**Review by darioush — APPROVED** [review]

File: scripts/install_avalanchego_release.sh
Comment (darioush) [reply scripts/install_avalanchego_release.sh:76]: Yes, however it also fetches other branches that are not needed for this run.
I think this is fine for now, in the longer term I think we should use github actions to [cache](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows) or create [artifacts](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts)

**Review by anusha-ctrl — APPROVED** [review]

**Pull request merged**
